File: MeepMeepWork/src/main/java/org/firstinspires/ftc/meepmeep/MeepMeepTesting.java
package org.firstinspires.ftc.meepmeep;

import com.acmerobotics.roadrunner.geometry.Pose2d;
import com.acmerobotics.roadrunner.geometry.Vector2d;
import com.noahbres.meepmeep.MeepMeep;
import com.noahbres.meepmeep.roadrunner.Constraints;
import com.noahbres.meepmeep.roadrunner.DefaultBotBuilder;
import com.noahbres.meepmeep.roadrunner.entity.RoadRunnerBotEntity;

import org.firstinspires.ftc.meepmeepkt.MeepMeepPersistence;
import org.firstinspires.ftc.teamcode.roadrunner.drive.DriveConstants;

public class MeepMeepTesting {
    static final String PERSISTENCE_FILE = "./MeepMeepWork/meepmeep.properties";

    public static void main(String[] args) {
        MeepMeep meepMeep = new MeepMeep(800);

        MeepMeepPersistence persistence = new MeepMeepPersistence(meepMeep, PERSISTENCE_FILE);
        persistence.restore();

        RoadRunnerBotEntity myBot = new DefaultBotBuilder(meepMeep)

            .setDimensions(15.5, 15)

            .setConstraints(new Constraints(
                DriveConstants.MAX_VEL,
                DriveConstants.MAX_ACCEL,
                DriveConstants.MAX_ANG_VEL,
                DriveConstants.MAX_ANG_ACCEL,
                DriveConstants.TRACK_WIDTH
            ))

            .followTrajectorySequence(drive ->
                drive.trajectorySequenceBuilder(new Pose2d(toInches(91), toInches(-159), radians(90)))
                    // Deposit preload
                    .splineTo(new Vector2d(toInches(91), toInches(-30)), radians(90))
                    .turn(radians(-135))

                    // Move to auto cycle and cycle first
                    .turn(radians(30))
                    .splineToSplineHeading(new Pose2d(toInches(140), toInches(-33), radians(0)), radians(0))

                    .setReversed(true)
                    .splineTo(new Vector2d(toInches(91), toInches(-30)), radians(160))
                    .setReversed(false)
                    .turn(radians(-25))

                    // 2nd auto cycle
                    .turn(radians(30))
                    .splineToSplineHeading(new Pose2d(toInches(140), toInches(-33), radians(0)), radians(0))

                    .setReversed(true)
                    .splineTo(new Vector2d(toInches(91), toInches(-30)), radians(160))
                    .setReversed(false)
                    .turn(radians(-25))

                    // 3rd auto cycle
                    .turn(radians(30))
                    .splineToSplineHeading(new Pose2d(toInches(140), toInches(-33), radians(0)), radians(0))

                    .setReversed(true)
                    .splineTo(new Vector2d(toInches(91), toInches(-30)), radians(160))
                    .setReversed(false)
                    .turn(radians(-25))


                    // 4th auto cycle
                    .turn(radians(30))
                    .splineToSplineHeading(new Pose2d(toInches(140), toInches(-33), radians(0)), radians(0))

                    .setReversed(true)
                    .splineTo(new Vector2d(toInches(91), toInches(-30)), radians(160))
                    .setReversed(false)
                    .turn(radians(-25))

                    // 5th auto cycle
                    .turn(radians(30))
                    .splineToSplineHeading(new Pose2d(toInches(140), toInches(-33), radians(0)), radians(0))

                    .setReversed(true)
                    .splineTo(new Vector2d(toInches(91), toInches(-30)), radians(160))
                    .setReversed(false)
                    .turn(radians(-70)) // Turn further to drive back to park

                    .splineTo(new Vector2d(toInches(91), toInches(-92)), radians(-90))
                    .strafeRight(toInches(60))

                    .build()
            );


        meepMeep.setBackground(MeepMeep.Background.FIELD_POWERPLAY_OFFICIAL)
            .setDarkMode(true)
            .setBackgroundAlpha(0.95f)
            .addEntity(myBot)
            .start();
    }

    public static double radians(double degrees) {
        return Math.toRadians(degrees);
    }

    // FROM CENTIMETERS
    public static double toInches(double centimeters) {
        return centimeters * 0.3837008;
    }
}

File: TeamCode/src/main/kotlin/org/firstinspires/ftc/teamcode/testing/MeepMeepTesting.kt
package org.firstinspires.ftc.teamcode.testing

import com.acmerobotics.roadrunner.geometry.Pose2d
import com.acmerobotics.roadrunner.geometry.Vector2d
import com.noahbres.meepmeep.MeepMeep
import com.noahbres.meepmeep.core.util.FieldUtil
import com.noahbres.meepmeep.roadrunner.DefaultBotBuilder
import org.firstinspires.ftc.teamcode.components.meepmeep.MeepMeepPersistence
import org.firstinspires.ftc.teamcode.roadrunner.drive.DriveConstants
import org.firstinspires.ftc.teamcode.components.meepmeep.MeepMeepFun
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.sin

const val PATH_RADIUS = 55
const val PATH_SIDES = 4
const val ROBOT_SIZE = 20
const val CONFIG_FILE_PATH =
    "TeamCode/src/main/res/raw/meepmeep.properties"

fun main() {
    val meepMeep = MeepMeep(1000, fps = 165)

    val persistence = MeepMeepPersistence(meepMeep, CONFIG_FILE_PATH)
    persistence.restore()

    val bot = DefaultBotBuilder(meepMeep)
        .setConstraints(
            maxVel = DriveConstants.MAX_VEL,
            maxAccel = DriveConstants.MAX_ACCEL,
            maxAngVel = DriveConstants.MAX_ANG_VEL,
            maxAngAccel = DriveConstants.MAX_ANG_ACCEL,
            trackWidth = DriveConstants.TRACK_WIDTH,
        )
        .followTrajectorySequence {
            data class Coordinate(val x: Double, val y: Double)

            val coords = (0 until PATH_SIDES).map { i ->
                val x = -PATH_RADIUS * sin(x = 2 * i * Math.PI / PATH_SIDES)
                val y = -PATH_RADIUS * cos(x = 2 * i * Math.PI / PATH_SIDES)
                Coordinate(x, y)
            }

            val startY = (ROBOT_SIZE - FieldUtil.FIELD_HEIGHT) / 2.0
            val startPose = Pose2d(0.0, startY, rad(90))

            it.trajectorySequenceBuilder(startPose).run {
                lineTo(Vector2d(coords[0].x, coords[0].y))

                for (i in 1..coords.size) {
                    val (x, y) = coords[i % coords.size]

                    val theta = atan2(rad(y), rad(x))
                    splineToLinearHeading(
                        Pose2d(x, y, theta + rad(180)),
                        theta
                    )

                    if (i < coords.size) {
                        addDisplacementMarker {}

                        UNSTABLE_addTemporalMarkerOffset(0.5) {}

                        waitSeconds(.51)
                    } else {
                        turn(Math.PI * 8)
                    }
                }

                lineTo(Vector2d(0.0, startY))
                build()
            }
        }

    MeepMeepFun(meepMeep, bot).launch()

    meepMeep
        .setBackground(MeepMeep.Background.GRID_GRAY)
        .setDarkMode(true)
        .addEntity(bot)
        .start()
}

fun rad(d: Int) = Math.toRadians(d.toDouble())
fun rad(d: Double) = Math.toRadians(d)